### PR TITLE
Ensure "latest" DockerHub image is identical to version-tagged image

### PR DESCRIPTION
**Pull request by mikejholly**
Addresses https://github.com/earthly/earthly/issues/2142

There's a new `BIN_VERSION` arg which can be manually set to differentiate between the binary version & the container image tag. It will fall back to the container image tag if it's not provided. 

Example:

```
$ earthly ./release+release-dockerhub --PUSH_LATEST_TAG=true --RELEASE_TAG="v8.8.8"
...
...
 Push Summary ⏫ (disabled)
————————————————————————————————————————————————————————————————————————————————

To enable pushing use earthly --push
Did not push image docker.io/earthly/buildkitd:v8.8.8
Did not push image docker.io/earthly/buildkitd:v8.8.8
Did not push image docker.io/earthly/buildkitd:v8.8.8
Did not push image docker.io/earthly/buildkitd:v8.8.8
Did not push image docker.io/earthly/buildkitd:latest
Did not push image docker.io/earthly/buildkitd:latest
Did not push image docker.io/earthly/buildkitd:latest
Did not push image docker.io/earthly/buildkitd:latest
Did not push image earthly/earthly:v8.8.8
Did not push image earthly/earthly:v8.8.8
Did not push image earthly/earthly:latest
Did not push image earthly/earthly:latest

 Local Output Summary 🎁
————————————————————————————————————————————————————————————————————————————————

Image +earthly-docker output as earthly/earthly:v8.8.8
Image +earthly-docker output as earthly/earthly:v8.8.8
Image +earthly-docker output as earthly/earthly:latest
Image +earthly-docker output as earthly/earthly:latest
Image ./buildkitd+buildkitd output as docker.io/earthly/buildkitd:v8.8.8
Image ./buildkitd+buildkitd output as docker.io/earthly/buildkitd:v8.8.8
Image ./buildkitd+buildkitd output as docker.io/earthly/buildkitd:latest
Image ./buildkitd+buildkitd output as docker.io/earthly/buildkitd:latest

```
And when checking the version string:

```
$ docker run --rm --entrypoint='' earthly/earthly:latest /usr/bin/earthly --version
earthly version v8.8.8 de7de9e0541b7478050dd3e50df6709a007be5bc linux/amd64; Alpine Linux 
$ docker run --rm --entrypoint='' earthly/earthly:v8.8.8 /usr/bin/earthly --version
earthly version v8.8.8 de7de9e0541b7478050dd3e50df6709a007be5bc linux/amd64; Alpine Linux
```

It doesn't seem like any doc changes are required as the new arg is specified in `release/Earthfile`.

### CODE REVIEW

**Comment by alexcb** [comment]
can you share the output of `docker images | grep -w 'earthly.earthly' | grep '\(latest\|v8.8.8\)'`?

I'm curious to see if the reported digests are the same.

**Comment by mikejholly** [comment]
```
docker images | grep -w 'earthly.earthly' | grep '\(latest\|v8.8.8\)'
earthly/earthly                                           v8.8.8_linux_arm64                                                       bbdcc479bb58   53 minutes ago      345MB
earthly/earthly                                           latest_linux_arm64                                                       0ae1a82c945c   53 minutes ago      345MB
earthly/earthly                                           latest                                                                   1707c3a779cb   About an hour ago   338MB
earthly/earthly                                           latest_linux_amd64                                                       1707c3a779cb   About an hour ago   338MB
earthly/earthly                                           v8.8.8                                                                   d0ea74f571ed   About an hour ago   338MB
earthly/earthly                                           v8.8.8_linux_amd64                                                       d0ea74f571ed   About an hour ago   338MB
```

**Comment by alexcb** [comment]
thanks for running that -- I think this needs more work, both `latest` and `v8.8.8` (or whatever actual release we're doing) should point to the same digest.

**Comment by mikejholly** [comment]
@alexcb Updated. 

I tried multiple `SAVE IMAGE`'s but the digests differed. A single `SAVE IMAGE` works but a dynamic image list is interpreted as a single quoted arg.

Example after running `earthly +release-dockerhub --PUSH_LATEST_TAG=true --RELEASE_TAG="v8.8.8"`

```
docker images | grep -w 'earthly/earthly'
earthly/earthly                                  latest_linux_arm64                         910ef350d26d   31 minutes ago   345MB
earthly/earthly                                  v8.8.8_linux_arm64                         910ef350d26d   31 minutes ago   345MB
earthly/earthly                                  latest                                     b6d636f8d528   31 minutes ago   338MB
earthly/earthly                                  latest_linux_amd64                         b6d636f8d528   31 minutes ago   338MB
earthly/earthly                                  v8.8.8                                     b6d636f8d528   31 minutes ago   338MB
earthly/earthly                                  v8.8.8_linux_amd64                         b6d636f8d528   31 minutes ago   338MB
```

It looks like the image IDs differ by arch. This seems to be expected based on my googling.

File: Earthfile [hunk 1]
```diff
@@ -451,17 +454,31 @@ earthly-all:
...
     ENV EARTHLY_IMAGE=true
     COPY earthly-entrypoint.sh /usr/bin/earthly-entrypoint.sh
     ENTRYPOINT ["/usr/bin/earthly-entrypoint.sh"]
     WORKDIR /workspace
-    COPY (+earthly/earthly --VERSION=$TAG --DEFAULT_INSTALLATION_NAME="earthly") /usr/bin/earthly
+    COPY (+earthly/earthly --TAG=$TAG --BIN_VERSION=$BIN_VERSION --DEFAULT_INSTALLATION_NAME="earthly") /usr/bin/earthly
     ARG DOCKERHUB_USER="earthly"
     ARG DOCKERHUB_IMG="earthly"
-    SAVE IMAGE --push --cache-from=earthly/earthly:main $DOCKERHUB_USER/$DOCKERHUB_IMG:$TAG
+    # Multiple SAVE IMAGE's lead to differing image digests, but multiple
```
Comment (alexcb) [inline Earthfile:470]: woohoo! nice find!

File: Earthfile [hunk 2]
```diff
@@ -341,22 +341,25 @@ earthly:
...
-    #     -GO_GCFLAGS='all=-N -l'
+    #     --GO_GCFLAGS='all=-N -l'
     ARG GO_GCFLAGS
     ARG EXECUTABLE_NAME="earthly"
     ARG DEFAULT_INSTALLATION_NAME="earthly-dev"
     RUN test -n "$GOOS" && test -n "$GOARCH"
     RUN test "$GOARCH" != "arm" || test -n "$VARIANT"
     ARG EARTHLY_TARGET_TAG_DOCKER
-    ARG VERSION="dev-$EARTHLY_TARGET_TAG_DOCKER"
+    # TAG is the container image tag whereas BIN_VERSION is the version
```
Comment (alexcb) [inline Earthfile:351]: is this change still needed? I don't easily follow why we need these values to be different.

Comment (mikejholly) [reply Earthfile:351]: Good point. I don't think it's needed with the `SAVE IMAGE` approach.

Comment (mikejholly) [reply Earthfile]: Updated. I also reverted `TAG` back to `VERSION` as it's more suitable for `+earthly`.

**Review by alexcb — APPROVED** [review]
